Fix clearing cache of InternalConfigScope

haampie opened the pull request:
Calling `clear()` on an InternalConfigScope object should do nothing, right now it permanently deletes the config.

haampie committed:
Add test for clear_caches for internal config scope

haampie committed:
InternalConfigScope clear() should be a no-op

alalazo commented on lib/spack/spack/test/config.py:
```diff
@@ -1132,3 +1132,14 @@ def test_single_file_scope_cache_clearing(env_yaml):
     assert before == after
+
+
+def test_internal_config_cache():
+    config = spack.config.Configuration()
+    config.push_scope(spack.config.InternalConfigScope('internal', {
+        'config': {
+            'build_jobs': 10
+        }
+    }))
+    config.clear_caches()
+    assert config.get('config:build_jobs') == 10
```
Can you open a related issue and:
```python
@pytest.mark.regression('XXXXX')
```
on top of this test?

alalazo commented on lib/spack/spack/test/config.py:
```diff
@@ -1132,3 +1132,14 @@ def test_single_file_scope_cache_clearing(env_yaml):
+            'build_jobs': 10
+        }
+    }))
+    config.clear_caches()
```
I would add a comment on what we're going to test for the next person reading this, for instance:
```suggestion
    # An InternalConfigScope object is constructed from data that is already
    # in memory, therefore it doesn't have any cache to clear. Here we ensure
    # that calling the clear method is a no-op.
    internal_scope.clear()
```

alalazo commented on lib/spack/spack/test/config.py:
```diff
@@ -1132,3 +1132,14 @@ def test_single_file_scope_cache_clearing(env_yaml):
+
+
+def test_internal_config_cache():
+    config = spack.config.Configuration()
+    config.push_scope(spack.config.InternalConfigScope('internal', {
```
Can we remove the external `Configuration` wrapper? I think the test would make sense already with just an `InternalConfigScope` that calls its `clear` method, is that correct?

haampie replied:
I don't know what the public api is for retrieving a value from a config scope object

haampie replied:
I don't want to write it like the tests above with

```
before = scope.get_section('config')
scope.clear()
after = scope.get_section('config')
assert before == after
```
because it's also true when `id(before) == id(after)` when `clear()` would mutate the config without rebinding it's internal config variable.

alalazo commented on lib/spack/spack/test/config.py:
```diff
@@ -1132,3 +1132,14 @@ def test_single_file_scope_cache_clearing(env_yaml):
     assert before == after
+
+
+def test_internal_config_cache():
```
I'd suggest:
```suggestion
def test_internal_config_scope_cache_clearing():
```
to be consistent with the test above

alalazo reviewed (changes requested):
Thanks for the quick bugfix! I have a few minor requests, let me know if anything is unclear.

haampie committed:
Incorporate review feedback

haampie committed:
rename test

alalazo commented on lib/spack/spack/test/config.py:
```diff
@@ -1132,3 +1132,20 @@ def test_single_file_scope_cache_clearing(env_yaml):
     assert before == after
+
+
+@pytest.mark.regression('22611')
+def test_internal_config_scope_cache_clearing():
+    """
+    An InternalConfigScope object is constructed from data that is already
+    in memory, therefore it doesn't have any cache to clear. Here we ensure
+    that calling the clear method is a no-op.
+    """
+    config = spack.config.Configuration()
+    config.push_scope(spack.config.InternalConfigScope('internal', {
+        'config': {
+            'build_jobs': 10
+        }
+    }))
+    config.clear_caches()
+    assert config.get('config:build_jobs') == 10
```
With respect to https://github.com/spack/spack/pull/22609/files#r603405923 what I am saying is that the bug doesn't need a `Configuration` object to be triggered. So something like:
```suggestion
@pytest.mark.regression('22611')
def test_internal_config_scope_cache_clearing():
    """
    An InternalConfigScope object is constructed from data that is already
    in memory, therefore it doesn't have any cache to clear. Here we ensure
    that calling the clear method is consistent with that..
    """
    data = {
        'config': {
            'build_jobs': 10
        }
    }
    internal_scope = spack.config.InternalConfigScope('internal', data)
    # Ensure that the initial object is properly set
    assert internal_scope.sections['config'] == data
    # Call the clear method
    internal_scope.clear()
    # Check that this didn't affect the scope object
    assert internal_scope.sections['config'] == data
```
would be removing any interaction with `Configuration` and still testing the bug.

haampie committed:
Update lib/spack/spack/test/config.py

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented:
It may be that fixing this bug revealed a test with issues

haampie committed:
Fix the tests

haampie commented:
Okay, the tests were relying on `spack.config.config.clear_caches()`, I've updated that to use a temporary config scope. Also I didn't like how the tests were testing the value of `spack.config.get('config:build_jobs')`, since that is only set when --jobs is present. So I've just split the tests to something more reasonable (imho).

haampie commented on lib/spack/spack/test/cmd/common/arguments.py:
```diff
@@ -16,53 +16,37 @@
 
-def test_negative_integers_not_allowed_for_parallel_jobs(parser):
+@pytest.mark.parametrize("ncores", [1, 2, 4, 8, 16, 32])
+def test_ommitted_job_flag(job_parser, ncores, monkeypatch):
```
```suggestion
def test_omitted_job_flag(job_parser, ncores, monkeypatch):
```

haampie committed:
Fix typo
